### PR TITLE
Custom domains: don't allow adding a custom domain on subprojects

### PR DESCRIPTION
**Pull request by stsewd**
Needs https://github.com/readthedocs/readthedocs.org/pull/8945/.

### CODE REVIEW

File: readthedocs/templates/projects/domain_list.html [hunk 1]
```diff
@@ -12,11 +12,22 @@
...
+        This project is a subproject of <a href="{{ superproject_url }}">{{ superproject }}</a>,
+        its documentation will always be served from the <code>{{ domain }}</code> domain.
+        <a href="https://docs.readthedocs.io/page/subprojects.html#custom-domain-on-subprojects">Learn more</a>.
+      {% endblocktrans %}
+    </p>
+  {% else %}
+    <p class="help_text">
+      {% blocktrans trimmed with docs_url='https://docs.readthedocs.io/page/custom_domains.html' %}
+          Configuring a custom domain allows you to serve your documentation from a domain other than "{{ default_domain }}". <a href="{{ docs_url }}">Learn more</a>.
+      {% endblocktrans %}
```
Comment (stsewd) [inline readthedocs/templates/projects/domain_list.html:28]: so, not sure if we should show both messages when this is a subproject.

File: readthedocs/templates/projects/domain_list.html [hunk 2]
```diff
@@ -12,11 +12,22 @@
...
 
 {% block project_edit_content %}
-  <p class="help_text">
-    {% blocktrans trimmed with docs_url='https://docs.readthedocs.io/page/custom_domains.html' %}
-        Configuring a custom domain allows you to serve your documentation from a domain other than "{{ default_domain }}". <a href="{{ docs_url }}">Learn more</a>.
-    {% endblocktrans %}
-  </p>
+  {% if superproject %}
+    {% url "projects_detail" superproject.slug as superproject_url %}
+    <p class="help_text">
```
Comment (stsewd) [inline readthedocs/templates/projects/domain_list.html:17]: don't think we have something like a class for a "warning" banner.

Comment (agjohnson) [reply readthedocs/templates/projects/domain_list.html:17]: We do not, not yet at least!

**Comment by agjohnson** [comment]
I haven't had a chance to pull this down locally to test, what does the UI look like with this addition? I'm assuming looking at the code that this doesn't alter the actually domain form UX, it just adds a warning block to the page.

I feel this is a good compromise, even though good UX here would probably be to block this scenario entirely and only allowing subprojects to remove existing domains. I assume if we dug into the data here, we'd only have a handful of projects in the particular state.

**Comment by agjohnson** [comment]
Actually, on second thought, is there a reason we don't want to do something like block the creation of new project custom domains if the project is a subproject?

**Comment by stsewd** [comment]
> Actually, on second thought, is there a reason we don't want to do something like block the creation of new project custom domains if the project is a subproject?

I think they are still useful as a "short link", subproject.domain.com/en/latest/foo.html -> main.domain.com/projects/subproject/en/latest/foo.html.

Maybe even to not break existing links if a project is being converted into a subproject (but the domain would be already there in most cases).

**Comment by stsewd** [comment]
It looks like this, a little opaque, but that's how it looks like right now.

Screenshot 2022-03-02 at 17-30-23 Edit Domains Read the Docs
Screenshot 2022-03-02 at 17-30-03 Edit Domains Read the Docs

Also, just saw that we are completely overriding this template on .com, so I'd need to port the changes or see if we can just use the same from .org (I think we should be able to do this now).

**Comment by stsewd** [comment]
> Also, just saw that we are completely overriding this template on .com, so I'd need to port the changes or see if we can just use the same from .org (I think we should be able to do this now).

Hmm, looks like the listing is better on .com, but it has specific classes that only exists on .com, so when porting those to .org it doesn't look right.

Screenshot 2022-03-02 at 18-07-13 Custom Domains - Read the Docs for Business

Screenshot 2022-03-02 at 18-07-45 Edit Domains Read the Docs

**Comment by humitos** [comment]
> I think they are still useful as a "short link", subproject.domain.com/en/latest/foo.html -> main.domain.com/projects/subproject/en/latest/foo.html.
> 
> Maybe even to not break existing links if a project is being converted into a subproject (but the domain would be already there in most cases).

If these are the only use cases we have in mind at the moment, I'd say to not allow creating a domain on a subproject as Anthony mentioned already, and explain that in the UI and documentation. I think allowing this just adds more confusion to our users and more complexity to our codebase with little winning.

**Comment by agjohnson** [comment]
Yeah, that's a good niche use case. But might be a bit confusing for most users, who would probably expect the subproject domain to take precedence or at least show in the request.

**Comment by stsewd** [comment]
Opened https://github.com/readthedocs/readthedocs.org/issues/8985 to discuss about removing that "feature". But if you all agree, then I can just change this PR to do that.

**Comment by agjohnson** [comment]
If you'd rather get this merged, that sounds fair. I will say I think the better user interaction will just be blocking domain creation. I've never been convinced that adding words to a UI makes any difference with what users do with the UI, so I'm skeptical that users will read this new block. But, no reason we can't split this into two chunks of work. I'm impartial on the order, whether we disable now and try to rework your use case back in, or we leave it enabled and eventually disable creation altogether.

**Comment by stsewd** [comment]
I have changed this PR so isn't allowed to create domains on subrprojects.

**Comment by ericholscher** [comment]
Looks like this is waiting on @agjohnson to review to move forward?

File: readthedocs/templates/projects/domain_list.html (hunk 1 above)
Comment (agjohnson) [reply readthedocs/templates/projects/domain_list.html:28]: I think it makes sense to just show one block. We have discussed on another issue that a better UX would be to show the form as disabled, but we can come back to this as well. There are maybe a few patterns to figure out there.

File: readthedocs/templates/projects/domain_list.html [hunk 3]
```diff
@@ -33,16 +44,18 @@ <h3> {% trans "Existing Domains" %} </h3>
...
-    {% include 'projects/includes/feature_disabled.html' with project=project %}
-  {% else %}
-    <form method="post" action="{% url 'projects_domains_create' project.slug %}">{% csrf_token %}
-      {{ form.as_p }}
-      <p>
-        <input style="display: inline;" type="submit" value="{% trans "Add" %}">
-      </p>
-    </form>
+  {% if not superproject %}
+    <h3>{% trans "Add new Domain" %}</h3>
```
Comment (agjohnson) [inline readthedocs/templates/projects/domain_list.html:48]: Might as well update while changing this:

```suggestion
    <h3>{% trans "Add new domain" %}</h3>
```

Based on style discussions on blog content and site content, headings deeper than `h1` can be sentence case.

**Review by agjohnson — APPROVED** [review]
This looks good :+1: I don't have any specific feedback with the approach in this PR now.

**Pull request merged**